### ISSUE TITLE
Adding more than 3 controllers breaks FastLED on an ESP32 using PlatformIO

### ISSUE THREAD

**Issue by nsamala** [FastLED/FastLED#722]
Using the Arduino framework for the ESP32 in PlatformIO , I have issues running code with more than 3 controllers. I'm using this with an ESP-WROOM32. It works fine for 3 controllers but the moment I add a fourth or fifth (see the `ledTest` function) it does not render the LEDs at all.

```c++
#include <Arduino.h>
#include <FastLED.h>
#include <vector>

const int STATUS_COUNT = 1;
const int STATUS_PIN = 21;
const int STRIP_COUNT = 45;
const int STRIP_ONE_PIN = 16;
const int STRIP_TWO_PIN = 17;
const int STRIP_THREE_PIN = 18;
const int STRIP_FOUR_PIN = 19;
TaskHandle_t ledTask;

void printInfo();

// led test methods
void ledTest(void *parameter);
void showGradient(CRGB *strip, uint8_t hue, int count);
void color(CRGB *strip, int count, CRGB color);
std::vector<CRGB*> strips;

void setup() {
    Serial.begin(115200);
    printInfo();

    xTaskCreate(ledTest, "ledTest", 5000, NULL, 1, &ledTask);
}

void loop() {
    delay(10);
}

void printInfo() {
    Serial.print("ESP SDK Version: ");
    Serial.println(ESP.getSdkVersion());

    Serial.print("ESP Chip Revision: ");
    Serial.println(ESP.getChipRevision());
}

void ledTest(void *parameter) {
    Serial.println("Starting LED test");
    strips.push_back(new CRGB[STATUS_COUNT]);
    strips.push_back(new CRGB[STRIP_COUNT]);
    strips.push_back(new CRGB[STRIP_COUNT]);
    strips.push_back(new CRGB[STRIP_COUNT]);
    strips.push_back(new CRGB[STRIP_COUNT]);

    FastLED.addLeds<WS2812B, STATUS_PIN, GRB>(strips[0], STATUS_COUNT);
    FastLED.addLeds<WS2812B, STRIP_ONE_PIN, GRB>(strips[1], STRIP_COUNT);
    FastLED.addLeds<WS2812B, STRIP_TWO_PIN, GRB>(strips[2], STRIP_COUNT);
    FastLED.addLeds<WS2812B, STRIP_THREE_PIN, GRB>(strips[3], STRIP_COUNT);
    FastLED.addLeds<WS2812B, STRIP_FOUR_PIN, GRB>(strips[4], STRIP_COUNT);

    int count;
    uint8_t hue = 0;
    for (;;) {
        hue = (hue + 1) % 255;
        for (int i = 0; i < strips.size(); i++) {
            count = i == 0 ? STATUS_COUNT : STRIP_COUNT;
            showGradient(strips[i], hue, count);
        }
        FastLED.show();
        delay(10);
    }
    vTaskDelete(NULL);
}

void showGradient(CRGB *strip, uint8_t hue, int count) {
    fill_rainbow(strip, count, hue);
}

void color(CRGB *strip, int count, CRGB color) {
    for (int i = 0; i != count; i++)
        (*strip)[i] = color;
}
```

**Comment by focalintent** [comment]
So - the code you have doesn't even compile - or at least it doesn't compile in the Arduino IDE (even if I fix the typo in your fill_rainbow line to ```fill_rainbow(strips, count, hue);``` - as there's no cast from std::vector<CRGB*> to CRGB* being found.

Can you try making this work under the Arduino IDE - if the 5 strips still isn't working there, I'll have @samguyer take a look at it - but platform IO isn't an officially tested/supported build environment.  I don't have any multi-strip esp32 setups here right now, and won't be able to spin one up for a bit, otherwise I'd test it here.

(Mostly I want to make sure that this is also an issue in the build environment that Sam is likely to be using before I send him off chasing)

**Comment by nsamala** [comment]
@focalintent sorry I had changed the code I had copied over in the sake of brevity and missed some things. I've gone ahead and fixed it. You should be able to run it in the Arduino IDE by removing the `#include <Arduino.h>` at the top.

When you comment out the last two `FastLED.addLED` lines, the strips iterate through the rainbow. Else, it freezes after setting that first status strip to red.

**Comment by focalintent** [comment]
Does it still not work for you will you build/install it through the Arduino IDE?

**Comment by focalintent** [comment]
Also does it work when you have the code in loop vs creating a task?

**Comment by nsamala** [comment]
@focalintent I've tried both. It doesn't work through the Arduino IDE and it doesn't work if I move the code to `loop()` instead of in a separate FreeRTOS task.

**Comment by nsamala** [comment]
@focalintent for reference, I've used the `SmartLeds` library [here](https://github.com/RoboticsBrno/SmartLeds) and it works for all five strips using RMT.

**Comment by focalintent** [comment]
Thanks - I’m trying to narrow the test case for @samguyer to look at.

**Comment by focalintent** [comment]
Also, @samguyer - pull the most recent build of the library - a) to see if there's any outstanding changes between what you have in your branch and what I have here and b) I went through and annotated the methods called from the interrupt handler with IRAM_ATTR

**Comment by nsamala** [comment]
@samguyer any luck reproducing this issue?

**Comment by kuike-tiel** [comment]
I have something like this, i am trying to make a led controller with the fastled library where the user can input wich led type they have and wich RGB arrangnement the led has. 
So i did try this and it worked.

```
void setFastled()
{
String lt = ledType;
 if (lt == "WS2812B") {
      ws2812B ();
 }  

void ws2812B ()
{
  String la = ledArrange;
    if (la == "RGB") {
     FastLED.addLeds<WS2812B, DATA_PIN, RGB>(leds, LED_NUMBER);
    } else if (la == "RBG") {
     FastLED.addLeds<WS2812B, DATA_PIN, RBG>(leds, LED_NUMBER);
    } else if (la == "GRB") {
    FastLED.addLeds<WS2812B, DATA_PIN, GRB>(leds, LED_NUMBER);
    } else if (la == "GBR") {
    FastLED.addLeds<WS2812B, DATA_PIN, GBR>(leds, LED_NUMBER);
    } else if (la == "BRG") {
     FastLED.addLeds<WS2812B, DATA_PIN, BRG>(leds, LED_NUMBER);
    } else if (la == "BGR") {
     FastLED.addLeds<WS2812B, DATA_PIN, BGR>(leds, LED_NUMBER);
    }
    FastLED.setMaxPowerInVoltsAndMilliamps(5, LED_LIMIT_MILLIAMPS);
  }


```

But when i do add a few more.
```
void setFastled()
{
String lt = ledType;
 
 if (lt == "WS2812B") {
      ws2812 ();
 }  
    if (lt == "WS2812") {
      ws2812 ();
    }
    if (lt == "WS2811") {
      ws2811 ();
    }
    if (lt == "WS2811_400") {
      ws2811_400 ();
    }
    if (lt == "SM16703") {
      sm16703 ();
    }
    if (lt == "TM1829") {
      tm1829 ();
    }
    if (lt == "TM1812") {
      tm1812 ();
    }
    if (lt == "TM1809") {
      tm1809 ();
    }
    if (lt == "TM1804") {
      tm1804 ();
    }
  }




void ws2812B ()
{
  String la = ledArrange;
    if (la == "RGB") {
     FastLED.addLeds<WS2812B, DATA_PIN, RGB>(leds, LED_NUMBER);
    } else if (la == "RBG") {
     FastLED.addLeds<WS2812B, DATA_PIN, RBG>(leds, LED_NUMBER);
    } else if (la == "GRB") {
    FastLED.addLeds<WS2812B, DATA_PIN, GRB>(leds, LED_NUMBER);
    } else if (la == "GBR") {
    FastLED.addLeds<WS2812B, DATA_PIN, GBR>(leds, LED_NUMBER);
    } else if (la == "BRG") {
     FastLED.addLeds<WS2812B, DATA_PIN, BRG>(leds, LED_NUMBER);
    } else if (la == "BGR") {
     FastLED.addLeds<WS2812B, DATA_PIN, BGR>(leds, LED_NUMBER);
    }
    FastLED.setMaxPowerInVoltsAndMilliamps(5, LED_LIMIT_MILLIAMPS);
  }
  
  void ws2812 ()
{
  String la = ledArrange;
    if (la == "RGB") {
     FastLED.addLeds<WS2812, DATA_PIN, RGB>(leds, LED_NUMBER);
    } else if (la == "RBG") {
     FastLED.addLeds<WS2812, DATA_PIN, RBG>(leds, LED_NUMBER);
    } else if (la == "GRB") {
    FastLED.addLeds<WS2812, DATA_PIN, GRB>(leds, LED_NUMBER);
    } else if (la == "GBR") {
    FastLED.addLeds<WS2812, DATA_PIN, GBR>(leds, LED_NUMBER);
    } else if (la == "BRG") {
     FastLED.addLeds<WS2812, DATA_PIN, BRG>(leds, LED_NUMBER);
    } else if (la == "BGR") {
     FastLED.addLeds<WS2812, DATA_PIN, BGR>(leds, LED_NUMBER);
    }
    FastLED.setMaxPowerInVoltsAndMilliamps(5, LED_LIMIT_MILLIAMPS);
  }


void ws2811()
{
  String la = ledArrange;
    if (la == "RGB") {
     FastLED.addLeds<WS2811, DATA_PIN, RGB>(leds, LED_NUMBER);
    } else if (la == "RBG") {
     FastLED.addLeds<WS2811, DATA_PIN, RBG>(leds, LED_NUMBER);
    } else if (la == "GRB") {
     FastLED.addLeds<WS2811, DATA_PIN, GRB>(leds, LED_NUMBER);
    } else if (la == "GBR") {
     FastLED.addLeds<WS2811, DATA_PIN, GBR>(leds, LED_NUMBER);
    } else if (la == "BRG") {
     FastLED.addLeds<WS2811, DATA_PIN, BRG>(leds, LED_NUMBER);
    } else if (la == "BGR") {
     FastLED.addLeds<WS2811, DATA_PIN, BGR>(leds, LED_NUMBER);
    }
    FastLED.setMaxPowerInVoltsAndMilliamps(5, LED_LIMIT_MILLIAMPS);
  }

void ws2811_400()
{
  String la = ledArrange;
    if (la == "RGB") {
     FastLED.addLeds<WS2811_400, DATA_PIN, RGB>(leds, LED_NUMBER);
    } else if (la == "RBG") {
     FastLED.addLeds<WS2811_400, DATA_PIN, RBG>(leds, LED_NUMBER);
    } else if (la == "GRB") {
     FastLED.addLeds<WS2811_400, DATA_PIN, GRB>(leds, LED_NUMBER);
    } else if (la == "GBR") {
     FastLED.addLeds<WS2811_400, DATA_PIN, GBR>(leds, LED_NUMBER);
    } else if (la == "BRG") {
     FastLED.addLeds<WS2811_400, DATA_PIN, BRG>(leds, LED_NUMBER);
    } else if (la == "BGR") {
     FastLED.addLeds<WS2811_400, DATA_PIN, BGR>(leds, LED_NUMBER);
    }
    FastLED.setMaxPowerInVoltsAndMilliamps(12, LED_LIMIT_MILLIAMPS);
}

void sm16703()
{
  String la = ledArrange;
    if (la == "RGB") {
     FastLED.addLeds<SM16703, DATA_PIN, RGB>(leds, LED_NUMBER);
    } else if (la == "RBG") {
     FastLED.addLeds<SM16703, DATA_PIN, RBG>(leds, LED_NUMBER);
    } else if (la == "GRB") {
     FastLED.addLeds<SM16703, DATA_PIN, GRB>(leds, LED_NUMBER);
    } else if (la == "GBR") {
     FastLED.addLeds<SM16703, DATA_PIN, GBR>(leds, LED_NUMBER);
    } else if (la == "BRG") {
     FastLED.addLeds<SM16703, DATA_PIN, BRG>(leds, LED_NUMBER);
    } else if (la == "BGR") {
     FastLED.addLeds<SM16703, DATA_PIN, BGR>(leds, LED_NUMBER);
    }
    FastLED.setMaxPowerInVoltsAndMilliamps(24, LED_LIMIT_MILLIAMPS);
  }


void tm1829()
{
  String la = ledArrange;
    if (la == "RGB") {
     FastLED.addLeds<TM1829, DATA_PIN, RGB>(leds, LED_NUMBER);
    } else if (la == "RBG") {
     FastLED.addLeds<TM1829, DATA_PIN, RBG>(leds, LED_NUMBER);
    } else if (la == "GRB") {
     FastLED.addLeds<TM1829, DATA_PIN, GRB>(leds, LED_NUMBER);
    } else if (la == "GBR") {
     FastLED.addLeds<TM1829, DATA_PIN, GBR>(leds, LED_NUMBER);
    } else if (la == "BRG") {
      FastLED.addLeds<TM1829, DATA_PIN, BRG>(leds, LED_NUMBER);
    } else if (la == "BGR") {
      FastLED.addLeds<TM1829, DATA_PIN, BGR>(leds, LED_NUMBER);
    }
    FastLED.setMaxPowerInVoltsAndMilliamps(5, LED_LIMIT_MILLIAMPS);
  }


void tm1812()
{
 String la = ledArrange;
    if (la == "RGB") {
      FastLED.addLeds<TM1812, DATA_PIN, RGB>(leds, LED_NUMBER);
    } else if (la == "RBG") {
      FastLED.addLeds<TM1812, DATA_PIN, RBG>(leds, LED_NUMBER);
    } else if (la == "GRB") {
      FastLED.addLeds<TM1812, DATA_PIN, GRB>(leds, LED_NUMBER);
    } else if (la == "GBR") {
      FastLED.addLeds<TM1812, DATA_PIN, GBR>(leds, LED_NUMBER);
    } else if (la == "BRG") {
      FastLED.addLeds<TM1812, DATA_PIN, BRG>(leds, LED_NUMBER);
    } else if (la == "BGR") {
      FastLED.addLeds<TM1812, DATA_PIN, BGR>(leds, LED_NUMBER);
    }
    FastLED.setMaxPowerInVoltsAndMilliamps(12, LED_LIMIT_MILLIAMPS);
  }
  void tm1809()
{
 String la = ledArrange;
    if (la == "RGB") {
      FastLED.addLeds<TM1809, DATA_PIN, RGB>(leds, LED_NUMBER);
    } else if (la == "RBG") {
      FastLED.addLeds<TM1809, DATA_PIN, RBG>(leds, LED_NUMBER);
    } else if (la == "GRB") {
      FastLED.addLeds<TM1809, DATA_PIN, GRB>(leds, LED_NUMBER);
    } else if (la == "GBR") {
      FastLED.addLeds<TM1809, DATA_PIN, GBR>(leds, LED_NUMBER);
    } else if (la == "BRG") {
      FastLED.addLeds<TM1809, DATA_PIN, BRG>(leds, LED_NUMBER);
    } else if (la == "BGR") {
      FastLED.addLeds<TM1809, DATA_PIN, BGR>(leds, LED_NUMBER);
    }
    FastLED.setMaxPowerInVoltsAndMilliamps(12, LED_LIMIT_MILLIAMPS);
  }
  void tm1804()
{
 String la = ledArrange;
    if (la == "RGB") {
      FastLED.addLeds<TM1804, DATA_PIN, RGB>(leds, LED_NUMBER);
    } else if (la == "RBG") {
      FastLED.addLeds<TM1804, DATA_PIN, RBG>(leds, LED_NUMBER);
    } else if (la == "GRB") {
      FastLED.addLeds<TM1804, DATA_PIN, GRB>(leds, LED_NUMBER);
    } else if (la == "GBR") {
      FastLED.addLeds<TM1804, DATA_PIN, GBR>(leds, LED_NUMBER);
    } else if (la == "BRG") {
      FastLED.addLeds<TM1804, DATA_PIN, BRG>(leds, LED_NUMBER);
    } else if (la == "BGR") {
      FastLED.addLeds<TM1804, DATA_PIN, BGR>(leds, LED_NUMBER);
    }
    FastLED.setMaxPowerInVoltsAndMilliamps(12, LED_LIMIT_MILLIAMPS);
  }

```

i get this error message:  

   section `.text' will not fit in region `iram1_0_seg'

 I am a noob and this is my first coding/project. So i do not know what i am doing wrong here. 
I tried this with a brand new sketch and also got the same error so it is not that my sketch is to big (my geuss).

**Comment by nsamala** [comment]
Closing this because it seems to be fixed in 3.2.9
